src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2836,8 +2834,31 @@ module Crystal
       false
     end
 
+    def visit(node : TypeRestriction)
+      obj = node.obj
+      to = node.to
+
+      obj.accept self
+
+      unless context = match_context
+        node.raise "BUG: there is no match context"
+      end
+
+      if type = obj.type.restrict(to, context)
+        node.type = type
+      else
+        node.raise "can't restrict #{obj.type} to #{to}"
```

Comment:
A spec is missing for this (the specs you added seem to be testing other things)